raft/handler.go

```diff
@@ -426,6 +424,19 @@ func (pm *ProtocolManager) startRaft() {
 				// the single call to `pm.applyNewChainHead` for more context.
 				lastAppliedIndex = lastPersistedCommittedIndex
 			}
+
+			// fix raft applied index out of range
+			firstIndex, err := pm.raftStorage.FirstIndex()
+			if err != nil {
+				panic(fmt.Sprintf("failed to read last persisted applied index from raft while restarting: %v", err))
+			}
+			lastPersistedAppliedIndex := firstIndex - 1
```

Comment:
Why `- 1` ? It seems that if we persisted the correct value earlier, then we'll overwrite it regardless.

Reply:
Hi @jpmsam 

This is as per the etcd raft implementation. `pm.raftStorage.FirstIndex()` returns `ms.ents[0].Index + 1` internally. Also, when raft library is checking if it is out of range, the lower boundary (included) is defined as `log.applied` which is set here https://github.com/jpmorganchase/quorum/blob/b7edc0b6c9316f4633270467902833e35ca9144b/vendor/github.com/coreos/etcd/raft/log.go#L65 Therefore, `lastPersistedAppliedIndex` is set to `firstIndex - 1` if it is out of sync.